client/state/user-settings/actions.js

```diff
@@ -0,0 +1,116 @@
+/**
+ * External dependencies
+ */
+import debugFactory from 'debug';
+
+/**
+ * Internal dependencies
+ */
+import {
+	USER_SETTINGS_REQUEST,
+	USER_SETTINGS_SAVE,
+	USER_SETTINGS_UPDATE,
+	USER_SETTINGS_UNSAVED_CLEAR,
+	USER_SETTINGS_UNSAVED_SET,
+	USER_SETTINGS_UNSAVED_REMOVE,
+} from 'state/action-types';
+
+const debug = debugFactory( 'calypso:user:settings' );
+
+/**
+ * Fetch user settings from WordPress.com API and store them in UserSettings instance
+ * @returns {Function} Action thunk
+ */
+export function fetchUserSettings() {
+	return {
+		type: USER_SETTINGS_REQUEST
+	};
+}
+
+/**
+ * Post settings to WordPress.com API at /me/settings endpoint
+ *
+ * @param {Object} settingsOverride - default settings object
+ * @return {Function} Action thunk
+ */
+export function saveUserSettings( settingsOverride ) {
+	return {
+		type: USER_SETTINGS_SAVE,
+		settingsOverride
+	};
+}
+
+/**
+ * Returns an action object signalling the settings have been received from server.
+ *
+ * @param  {Object} settingValues Setting values (the subset of keys to be updated)
+ * @return {Object}               Action object
+ */
+export function updateUserSettings( settingValues ) {
+	return {
+		type: USER_SETTINGS_UPDATE,
+		settingValues
+	};
+}
+
+export function cancelPendingEmailChange() {
+	return {
+		type: USER_SETTINGS_SAVE,
+		settingsOverride: { user_email_change_pending: false }
+	};
+}
+
+/**
+ * Handles the storage and removal of changed setting that are pending
+ * being saved to the WPCOM API.
+ *
+ * @param  {String} settingName - setting name
+ * @param  {*} value - setting value
+ * @return {Function} Action thunk that returns updating successful response
+ */
+export function updateUserSetting( settingName, value ) {
+	return ( dispatch, getState ) => {
+		const { settings } = getState().userSettings;
+
+		if ( ! settings || settings[ settingName ] === undefined ) {
+			debug( settingName + ' does not exist in user-settings data module.' );
+			return false;
+		}
+
+		/*
+		 * If the two match, we don't consider the setting "changed".
+		 * user_login is a special case since the logic for validating and saving a username
+		 * is more complicated.
+		 */
+		if ( settings[ settingName ] === value && 'user_login' !== settingName ) {
+			debug( 'Removing ' + settingName + ' from changed settings.' );
+			dispatch( removeUnsavedUserSetting( settingName ) );
+		} else {
+			dispatch( updateUnsavedUserSetting( settingName, value ) );
+		}
+
+		return true;
+	};
```

Comment:
it actually looks like this is only called from within the middleware, however, and making it a thunk has only moved the statefulness out of the middleware and into the action creator. this is a pattern I have a hard time accepting.

usually when I think that we need thunks or stateful action creators I recognize that I haven't fully considered alternatives often in the data structure itself

in this case since it's triggered from the middleware we have an easy way out - just pass in the old `settings` as an additional parameter

on the other hand, why do we have a specific `removeUnsavedUserSetting` action? do we actually need that or does the act of updating it imply that we can remove it? if the value is the same then an update should be idempotent and no real harm is done by going ahead and updating it. in this scenario we not only don't need to pass in the `settings` object and this becomes stateless, but we can have the unsaved settings reducer watch for the update action and remove the dirty bit

anyway, I'm brainstorming here because I don't know the behavioral goals as well as you do. introducing a new thunk is something I hope we can avoid entirely